pkg/app/api/grpcapi/piped_api.go

```diff
@@ -1105,6 +1105,56 @@ func (a *PipedAPI) CreateDeploymentChain(ctx context.Context, req *pipedservice.
 	return &pipedservice.CreateDeploymentChainResponse{}, nil
 }
 
+// InChainDeploymentPlannable hecks the completion and status of the previous block in the deployment chain.
+// An in chain deployment is treated as plannable in case:
+// - It's the first deployment of its deployment chain.
+// - All deployments of its previous block in chain are at DEPLOYMENT_SUCCESS state.
+func (a *PipedAPI) InChainDeploymentPlannable(ctx context.Context, req *pipedservice.InChainDeploymentPlannableRequest) (*pipedservice.InChainDeploymentPlannableResponse, error) {
+	_, pipedID, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+	if err := a.validateDeploymentBelongsToPiped(ctx, req.Deployment.Id, pipedID); err != nil {
+		return nil, err
+	}
+
+	dc, err := a.deploymentChainStore.GetDeploymentChain(ctx, req.Deployment.DeploymentChainId)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "unable to find the deployment chain which this deployment belongs to")
+	}
+
+	// Deployment of blocks[0] in the chain means it's the first deployment of the chain;
+	// hence it should be processed without any lock.
+	if req.Deployment.DeploymentChainBlockIndex == 0 {
+		return &pipedservice.InChainDeploymentPlannableResponse{
+			Plannable: true,
+		}, nil
+	}
+
+	if req.Deployment.DeploymentChainBlockIndex >= int32(len(dc.Blocks)) {
+		return nil, status.Error(codes.InvalidArgument, "invalid deployment with chain block index provided")
+	}
+
+	prevBlock := dc.Blocks[req.Deployment.DeploymentChainBlockIndex-1]
```

Comment:
Since `DeploymentChainBlockIndex` is `int32` which is signed integer whose range is -2147483648 to 2147483647, this is a bit likely to be panic.

How about making `DeploymentChainBlockIndex` an unsigned integer, or adding a simple validation?

Reply:
Nice catch, golang does not support negative indexes for slide, right? In that case, lets me address it in a separated PR since other place that not related to this change need to be updated too 👍

Reply:
Fine to me 👍